### PR TITLE
[openai] No side effects when streaming; stay resilient to edge cases with parse(custom_type)

### PR DESCRIPTION
**Pull request by lmolkova**
Replaces https://github.com/open-telemetry/opentelemetry-python-genai/pull/147

Today we call `parse` on legacy streams introducing side-effects (changing stream type for end users). This leads to issues like #46 (not having methods / properties that stream has without instrumentation being applied).

This PR uses different approach: 

- Wrap with_raw_response streaming transparently - parse deferred until/if user calls it
- Finalize the span from an httpx close/aclose fallback for callers that never parse. We won't record content / response attributes, but still end the span. 
- Tolerate caller-defined chunk/event types: skip chunks with unknown type

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    ) -> None:
+        super().__init__(raw_response)
+        self._self_wrap_stream = wrap_stream
+        self._self_finalize: Callable[[], None] | None = finalize
+        self._self_parsed: object | None = None
+        self._install_close_fallback(raw_response)
+
+    def _install_close_fallback(self, raw_response: object) -> None:
+        # httpx exposes a sync ``close`` and an async ``aclose``; wrap whichever
+        # the underlying response has so the appropriate one finalizes the span.
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:95]: Wondering does this play nice with `opentelemetry-instrumentation-httpx`?

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    ends by closing the underlying httpx response — so a fallback on its
+    ``close``/``aclose`` finalizes the span when ``parse()`` never built a
+    wrapper that would finalize it instead.
+    """
+
+    def __init__(
+        self,
+        raw_response: object,
+        wrap_stream: Callable[[object], object],
+        finalize: Callable[[], None],
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py]: Is it possible to type these more tightly than `object`?

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    def _install_close_fallback(self, raw_response: object) -> None:
+        # httpx exposes a sync ``close`` and an async ``aclose``; wrap whichever
+        # the underlying response has so the appropriate one finalizes the span.
+        http_response = getattr(raw_response, "http_response", None)
+        close = getattr(http_response, "close", None)
+        if close is not None:
+            http_response.close = self._wrap_sync_close(close)
+        aclose = getattr(http_response, "aclose", None)
+        if aclose is not None:
+            http_response.aclose = self._wrap_async_close(aclose)
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py]: Going along with above comment about typing the ctor inputs a little more, can we use `isinstance()` checks or some other type narrowing strategy?

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        # When ``parse()`` built a stream wrapper, that wrapper owns
+        # finalization; only finalize here for callers that never parsed.
+        if self._self_parsed is None:
+            self._finalize_once()
+
+    def _finalize_once(self) -> None:
+        # ``stop()`` is not idempotent, so finalize at most once.
+        if self._self_finalize is not None:
+            finalize, self._self_finalize = self._self_finalize, None
+            finalize()
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:138]: do we need it to be thread safe?

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+    def parse(self, *args: object, **kwargs: object) -> object:
+        # We memoize the first parse regardless of the arguments it was called
+        # with, so calling parse() again with different args (e.g. a different
+        # ``to=`` type) returns the same wrapper rather than re-parsing. This is
+        # deliberate: one raw response backs one span, and re-parsing a stream
+        # would consume it twice anyway.
+        if self._self_parsed is not None:
+            return self._self_parsed
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:147]: also not thread safe but maybe that's ok

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    """
+
+    @classmethod
+    def wrap_result(
+        cls: type,
+        result: object,
+        invocation: GenAIInvocation,
+        capture_content: bool,
+    ) -> object:
+        if hasattr(result, "parse"):
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py]: might be reasonable to just check for `isinstance(result, LegacyAPIResponse)` since you mentioned it may change in future versions and we can narrow down the code here.

alternatively if you want it to be open to changes or we end up supporting multiple versions, I'm partial to using Protocols with [`@runtime_checkable`](https://typing.python.org/en/latest/spec/protocol.html#runtime-checkable-decorator-and-narrowing-types-by-isinstance) to capture intent in the typing but maybe it's too annoying

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    The concrete wrapper must accept ``(stream, invocation, capture_content)``.
+    Call it from a streaming branch: a ``with_raw_response`` result (a
+    ``LegacyAPIResponse``, detected by its ``parse`` method) is wrapped so its
+    metadata resolves natively and ``parse()`` is deferred; a plain SDK stream
+    is wrapped directly.
+    """
+
+    @classmethod
+    def wrap_result(
+        cls: type,
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py]: nit maybe a stylistic thing, but im not personally a fan of mixins like this because it silently depends on all constructors having a compatible signature

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/response_wrappers.py [hunk 8]
```diff
@@ -178,13 +181,24 @@ def response(self):
         return _ResponseProxy(response, lambda: self._stop(None))
 
     def process_event(self, event: "ResponseStreamEvent[TextFormatT]") -> None:
-        event_type = event.type
+        # raw-response stream can be parsed into a caller-defined event type.
+        event_type = getattr(event, "type", None)
+        if not isinstance(event_type, str):
+            # Log once per stream to avoid spamming on long streams.
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/response_wrappers.py]: Since it's debug level, maybe it's OK to just spam? Idk keep me honest but I'm not sure it's worth the effort

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/response_wrappers.py [hunk 9]
```diff
@@ -178,13 +181,24 @@ def response(self):
...
+                )
+            return
+
         response: "ParsedResponse[TextFormatT] | Response | None" = getattr(
             event, "response", None
         )
 
         if response and not self._self_invocation.request_model:
-            model = response.model
+            model = getattr(response, "model", None)
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/response_wrappers.py:194]: why do we need this change? Judging by the "cast" on L196, I think it's OK

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+"""Transparent proxy for streaming ``with_raw_response`` results."""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING, Awaitable, Callable
+
+from openai import AsyncStream, Stream
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:21]: Does this cause an `ImportError` if the instrumenting is attempted but the OpenAI SDK is not installed or not a compatible version? I think that was the reason for the lazy/defensive imports we previously had, but it might be OK already depending if this module is imported lazily from the `Instrumentor`

File: .github/instructions/instrumentation.instructions.md [hunk 11]
```diff
@@ -36,6 +36,11 @@ prefer opt-in or additive. Breaking changes need explicit justification in the P
 - Adding attributes to invocations produced by the util is fine.
 - Streaming responses must be instrumented by subclassing the util's `SyncStreamWrapper` /
   `AsyncStreamWrapper` (`opentelemetry.util.genai.stream`). Flag hand-rolled stream wrappers.
+- Instrumentation must not change what a call returns or when its work happens. Flag: side effects
```
Comment (aabmass) [inline .github/instructions/instrumentation.instructions.md]: My 2c, there will always be some whack-a-mole, but we gotta keep our sanity in terms of limiting side effects or anticipating future changes. I hope this doesn't make the AI take it too far 😆 

As an extreme example (that I think is tolerable risk), some of the function wrappers added like `_wrap_async_close()` could still break users if 

- the callable is a class implementing `__call__()` (pretty common in python but not the case for HTTPX)
- user is inspecting type signatures, docstrings, etc

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                return original(*args, **kwargs)
+            finally:
+                self._finalize_close_fallback()
+
+        return _close
+
+    def _wrap_async_close(
+        self, original: Callable[..., Awaitable[object]]
+    ) -> Callable[..., Awaitable[object]]:
+        async def _aclose(*args: object, **kwargs: object) -> object:
```
Comment (aabmass) [inline instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:120]: consider using https://docs.python.org/3/library/functools.html#functools.wraps

**Review by aabmass — APPROVED** [review]
Sorry for the nitpicky comments on typing. Approach LGTM.

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 6 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py]: LegacyAPIResponse is in internal module and there are a few more types this thing can be, added Protocols with @runtime_checkable

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 7 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py]: replaced with module function - looks much cleaner

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/response_wrappers.py (hunk 9 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/response_wrappers.py:194]: I believe it came up in some test  - it's needed because it breaks in corner cases

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 10 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:21]: we're testing with oldest version of openai SDK, so it should be ok, also I don't believe we protect anywhere from original package not being installed

File: .github/instructions/instrumentation.instructions.md (hunk 11 above)
Comment (lmolkova) [reply .github/instructions/instrumentation.instructions.md]: yeah, soften it

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 12 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:120]: nice, thanks!

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 5 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:147]: response streams we're wrapping are not thread safe so it should be ok

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 4 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:138]: made underlying utils idempotent to multiple calls, but thread safety should not be necessary - underlying objects aren't

File: instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py (hunk 1 above)
Comment (lmolkova) [reply instrumentation/opentelemetry-instrumentation-genai-openai/src/opentelemetry/instrumentation/genai/openai/_raw_response.py:95]: it does, but added test. httpx instrumentation patches different methods - we don't await stream closure in http because it's super-fragile there

**Review by eternalcuriouslearner — APPROVED** [review]
